From e63374f347ceedecc3c821022f2ca587db47c529 Mon Sep 17 00:00:00 2001
From: Dmitry Osipenko <dmitry.osipenko@collabora.com>
Date: Tue, 14 Oct 2025 07:07:46 +0300
Subject: [PATCH 097/188] drm/fbdev-client: Don't enable fbdev emulation for
 virtio-gpu driver

VirtIO-GPU driver has a known problem with fbdev emulation + native context
where guest Xorg would create framebuffer from a dumb (vrend) BO before
native context has been set up for the Xorg's DRM context. This results in
a corrupted rendering when that dumb BO is further used from within native
context, which rejects it and skips rendering. Problem affects only Xorg and
not presents with Wayland. The fix primary wanted for SDDM and other DMs that
use Xorg.

Signed-off-by: Dmitry Osipenko <dmitry.osipenko@collabora.com>
---
 drivers/gpu/drm/clients/drm_client_setup.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/clients/drm_client_setup.c b/drivers/gpu/drm/clients/drm_client_setup.c
index 515aceac2..37d4549fe 100644
--- a/drivers/gpu/drm/clients/drm_client_setup.c
+++ b/drivers/gpu/drm/clients/drm_client_setup.c
@@ -40,7 +40,7 @@ void drm_client_setup(struct drm_device *dev, const struct drm_format_info *form
 	}
 
 #ifdef CONFIG_DRM_FBDEV_EMULATION
-	if (!strcmp(drm_client_default, "fbdev")) {
+	if (strcmp(dev->driver->name, "virtio_gpu") && !strcmp(drm_client_default, "fbdev")) {
 		int ret;
 
 		ret = drm_fbdev_client_setup(dev, format);
-- 
2.55.0

